library/htmlspecialchars.inc.php

```diff
@@ -55,13 +55,25 @@ function errorLogEscape($text)
  * Escape variables that are outputted into csv and spreadsheet files.
  * See here: https://www.owasp.org/index.php/CSV_Injection
  * Based mitigation strategy on this report: https://asecurityz.blogspot.com/2017/12/csv-injection-mitigations.html
- *  1. Remove all the following characters:  = + - @ " |
- *  2. Prepend value with a space.
- *  3. Remove trailing spaces (this specific strategy was reported here: https://www.contextis.com/en/blog/comma-separated-vulnerabilities).
+ *  1. Remove all the following characters:  = + " |
+ *  2. Then remove any leading - characters (not removing all of these since need in dates)
+ *  3. Then remove all leading @ characters (not removing all of these since need in email addresses)
+ *  4. Prepend value with a space.
+ *  5. Remove trailing spaces (this specific strategy was reported here: https://www.contextis.com/en/blog/comma-separated-vulnerabilities).
+ * If needed in future, will add a second parameter called 'mode' that will allow less stringent (or more stringent) mechanisms to escape for csv.
  */
 function csvEscape($text)
 {
-    return " " . rtrim(preg_replace('/[=+-@"|]/', '', $text));
+    // 1. Remove all the following characters:  = + " |
+    $text = preg_replace('/[=+"|]/', '', $text);
+
+    // 2. Then remove any leading - characters (not removing all of these since need in dates)
+    // 3. Then remove all leading @ characters (not removing all of these since need in email addresses)
+    $text = preg_replace('/^[\-@]+/', '', $text);
```

Comment:
any = all :)

Reply:
clarified documentation. hopefully it's more clear now :)

Reply:
I've been following and don't understand. Best to scan for vulnerability and reject document rather than to try to fix...No?

Reply:
looks great, nice use of preg_replace, the ^, the escape \ of the - and the + for more than 1 - or @

Reply:
Can't you just put the whole string in double quotes to ensure it's interpreted as a literal and not as a formula?

Reply:
Answering my own question after a simple test... with LibreOffice if you do that AND enable "Format quoted field as text" at import time then it will not treat it as a formula. So maybe not good enough.

By the way has anyone looked at the PhpSpreadsheet library? Seems it can generate Excel (and other formats) and offer better control.

https://github.com/PHPOffice/PhpSpreadsheet

Reply:
@sjpadgett and @sunsetsystems ,
yep, this is a rather not at all straightforward vulnerability to address. There is not a simple one size fits all fix or check to run here. The vulnerability fix really is dependent on what is being stored in the csv files and what the csv files are going to be used for. So need to balance potentially breaking things with the fix. The current mechanism should suffice well for patient demographics, but could be overkill in cases where accounting is happening (when we run into that, we can further refine this functionality). Whenever I think we are done with escaping functions, I then get an email with a vulnerability to create another one... :)

Reply:
Okay, instead of trying to have a one escape fit all then really the escaping should happen in the import routines.  We know what we want in each imported field, check the data types there. Should alleviate snooping entire file and trying to predict if the import will work everywhere. 
Still I may be missing something but too me, if a flat file is misformatted for the import we are trying to do then send it back to user as a suspect or unexpected format.

Reply:
Sanitizing input in OpenEMR would be a good thing to tackle, but would be a large endeavor.

Just incorporated this function in rest of codebase to get a good idea of all the use cases. And changed this function a bit to the following:
 1. Remove all the following characters:  = + " |
 2. Only remove leading - characters (since need in dates)
 3. Only remove leading @ characters (since need in email addresses)
 4. Surround with double quotes (no reference link, but seems very reasonable, which will prevent commas from breaking things).

These steps should:
1. Not break anything since csv outputs in OpenEMR are just holding standard data and not doing anything fancy.
2. Should make it impossible to bring in CSV injection attacks

And if we run into use cases that require specialized processing, then can address it by adding a second `$options` parameter to the function (I honestly don't think we will ever need to do this, but just saying there is room for growth).

Reply:
Removing leading minus sign prevents export of negative numbers. With the collections report it is important to know if the balance is a credit or not. I'd like that `$options` parameter, if you had an idea of how it would be implemented.

Reply:
hi @surfacehazard how about something like
```
* @param string $text The string to escape, possibly including "=", "+", or "-".
* @param array $options The optional array of calling report to exclude from default filter.
* @return string The string, with "=", "+", and "-" escaped dependent on optional $options.
*/
function csvEscape($text, $options = [])
{
    // 1. Remove all the following characters:  = + " |
    $text = preg_replace('/[=+"|]/', '', $text);

    // 2. Only remove leading - characters (since need in dates)
    // 3. Only remove leading @ characters (since need in email addresses)
    if (in_array('collections_report', $options)) {
        $text = preg_replace('/^[\@]+/', '', $text);
    } else {
        $text = preg_replace('/^[\-@]+/', '', $text);
    }

    // 4. Surround with double quotes (no reference link, but seems very reasonable, which will prevent commas from breaking things).
    return '"' . $text . '"';
}
```
and then call it from the billing_report with
```
                echo csvEscape(oeFormatMoney($balance), $options = ['collections_report'])              . ',';
```

Reply:
Yes, that's great! Left up to me, I think I would make the option more generic, like "export_csv_negative_numbers", such that it would make more sense in a config GUI.

Reply:
Agree on making it more generic, and when do this exception would put a single quote in front of the - character (hopefully it works on testing, if not, then can regroup).
source:
https://www.comparitech.com/blog/information-security/csv-injection-attacks/
"
If your application produces CSV files, you can perform whitelist validation on untrusted input and disallow the Equals (=), Plus (+), Minus (-), and At (@) characters. Whitelist validation simply means creating a whitelist of allowed characters and referencing input against the whitelist.  Any characters not on the whitelist are disallowed and removed. This is probably the safest method. However, it assumes your web site/application doesn’t need to allow these characters in order to perform its functions. 

If it does need to accept those characters, you can encode cell values so that the CSV application won’t treat these characters as formulae by preceding cell values that begin with the characters: =, +, -, or @ with a single quote. This method is referred to as “escaping” the characters and ensures that these characters will be interpreted as data rather than as formulae.
"